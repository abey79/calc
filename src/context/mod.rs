pub mod ast;
pub mod checked_ast;
pub mod source;
pub mod token_stream;
