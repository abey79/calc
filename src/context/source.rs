pub struct Source(String);

impl Source {
    pub fn new(source: String) -> Self {
        Self(source)
    }

    pub fn source(&self) -> &str {
        &self.0
    }
}
