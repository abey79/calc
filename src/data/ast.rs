//! This module contains the AST data structures.
//!
//! Throughout the code, the "Node" terminology is used to refer to elements of the AST. AST nodes
//! are assigned a unique `NodeId` when created. Factory functions are also provided to make it easy
//! to build AST nodes.

use crate::data::meta::Meta;
use crate::data::token::TokenKind;
use std::fmt;
use std::fmt::{Debug, Display};

// =================================================================================================
// DATA STRUCTURES

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOpKind {
    Pos,
    Neg,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind<T: Debug + Display> {
    Variable(VarName<T>),
    BinOp {
        op: BinOp<T>,
        left: Box<Expr<T>>,
        right: Box<Expr<T>>,
    },
    UnaryOp {
        op: UnaryOp<T>,
        operand: Box<Expr<T>>,
    },
    Tuple(Vec<Expr<T>>),
    Integer(i32),
    Float(f64),
}

#[derive(Debug, Clone, PartialEq)]
pub enum StmtKind<T: Debug + Display> {
    Assign { name: VarName<T>, value: Expr<T> },
    Print { expr: Expr<T> },
    Expr { expr: Expr<T> },
}

pub type BinOp<T> = Meta<BinOpKind, T>;
pub type UnaryOp<T> = Meta<UnaryOpKind, T>;
pub type VarName<T> = Meta<String, T>;
pub type Expr<T> = Meta<ExprKind<T>, T>;
pub type Stmt<T> = Meta<StmtKind<T>, T>;

// =================================================================================================
// FACTORIES

impl From<&TokenKind> for UnaryOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Pos,
            TokenKind::Minus => Self::Neg,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl<T> UnaryOp<T> {
    pub fn new(kind: impl Into<UnaryOpKind>, meta: impl Into<T>) -> Self {
        Self {
            kind: kind.into(),
            meta: meta.into(),
        }
    }
}

impl From<&TokenKind> for BinOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Add,
            TokenKind::Minus => Self::Sub,
            TokenKind::Star => Self::Mul,
            TokenKind::Slash => Self::Div,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl<T> BinOp<T> {
    pub fn new(kind: impl Into<BinOpKind>, meta: impl Into<T>) -> Self {
        Self {
            kind: kind.into(),
            meta: meta.into(),
        }
    }
}

impl<T> VarName<T> {
    pub fn new(name: impl Into<String>, meta: impl Into<T>) -> Self {
        Self {
            kind: name.into(),
            meta: meta.into(),
        }
    }
}

impl<T: Debug + Display> Expr<T> {
    pub fn variable(name: impl Into<VarName<T>>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Variable(name.into()),
            meta: meta.into(),
        }
    }

    pub fn bin_op(op: BinOp<T>, left: Expr<T>, right: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::BinOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            },
            meta: meta.into(),
        }
    }

    pub fn unary_op(op: UnaryOp<T>, operand: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::UnaryOp {
                op,
                operand: Box::new(operand),
            },
            meta: meta.into(),
        }
    }

    pub fn tuple(exprs: Vec<Expr<T>>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Tuple(exprs),
            meta: meta.into(),
        }
    }

    pub fn integer(value: i32, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Integer(value),
            meta: meta.into(),
        }
    }

    pub fn float(value: f64, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Float(value),
            meta: meta.into(),
        }
    }
}

impl<T: Debug + Display> Stmt<T> {
    pub fn assign(name: impl Into<VarName<T>>, value: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: StmtKind::Assign {
                name: name.into(),
                value,
            },
            meta: meta.into(),
        }
    }

    pub fn print(expr: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: StmtKind::Print { expr },
            meta: meta.into(),
        }
    }

    pub fn expr(expr: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: StmtKind::Expr { expr },
            meta: meta.into(),
        }
    }
}

// =================================================================================================
// UTILITY TRAITS

impl fmt::Display for BinOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
        }
    }
}

impl fmt::Display for UnaryOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Pos => write!(f, "+"),
        }
    }
}
