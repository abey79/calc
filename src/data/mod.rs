pub mod ast;
pub mod identified;
pub mod span;
pub mod token;
