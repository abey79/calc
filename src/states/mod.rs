//! This module models states (as in "state machine") between the compilation steps.
//!
//! The data associated with each state is factored in "context" structure for reusability (e.g. the
//! [`TokenStream`] context ("state data") is used in both the [`TokenizedState`] and the
//! [`ParsedState`]).

pub mod input_state;
pub mod parsed_state;
pub mod tokenized_state;

pub use input_state::*;
pub use parsed_state::*;
pub use tokenized_state::*;
