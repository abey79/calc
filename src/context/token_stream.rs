//! This context stores the tokens and their spans.

use crate::data::span::Span;
use crate::data::token::{Token, TokenId};
use crate::data::token_span::TokSpan;
use std::collections::BTreeMap;
use std::rc::Rc;

#[derive(Debug, Default)]
pub struct TokenStream {
    tokens: Vec<Rc<Token>>,
    tokens_by_id: BTreeMap<TokenId, Rc<Token>>,
    token_spans: BTreeMap<TokenId, Span>,
}

impl TokenStream {
    pub fn tokens(&self) -> &[Rc<Token>] {
        &self.tokens
    }

    pub fn token_from_id(&self, id: TokenId) -> Option<&Token> {
        self.tokens_by_id.get(&id).map(|t| &**t)
    }

    pub fn span_from_id(&self, id: TokenId) -> Option<Span> {
        self.token_spans.get(&id).copied()
    }

    pub fn push_token(&mut self, token: Token, span: Span) {
        self.token_spans.insert(token.id, span);
        let token = Rc::new(token);
        self.tokens.push(token.clone());
        self.tokens_by_id.insert(token.id, token);
    }

    pub fn span_from_tok_span(&self, tok_span: TokSpan) -> Option<Span> {
        let start = self.span_from_id(tok_span.start)?;
        let end = self.span_from_id(tok_span.end)?;
        Some(Span::new(start.start, end.end))
    }
}
