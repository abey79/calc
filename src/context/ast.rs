use crate::data::ast::Stmt;
use crate::data::meta::Meta;
use crate::data::span::Span;
use crate::data::token_span::TokSpan;
use crate::pipeline;
use std::fmt;
use std::fmt::{Debug, Display, Write};

#[derive(Debug, Default)]
pub struct Ast<M: Debug + Display> {
    stmts: Vec<Stmt<M>>,
}

impl<M: Debug + Display> Ast<M> {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[Stmt<M>] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<Stmt<M>> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: Stmt<M>) {
        self.stmts.push(stmt);
    }

    pub fn format<W: fmt::Write>(&self, w: &mut W) -> Result<String, fmt::Error> {
        pipeline::formatter::format(self, w)
    }
}

impl Ast<TokSpan> {
    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in self.stmts() {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Provide convenient methods for AST nodes parametrized by a token span.
impl<K> Meta<K, TokSpan> {
    pub fn tok_span(&self) -> TokSpan {
        self.meta.clone()
    }

    pub fn span(&self) -> Span {
        self.meta.span()
    }
}
