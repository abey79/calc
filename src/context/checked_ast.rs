//! State after the type checker.
//!
//! In this state, all the AST nodes are decorated with type info.

use crate::data::ast::{Expr, Stmt};
use crate::data::token_span::TokSpan;
use std::fmt;
use std::fmt::Write;

pub type CheckedStmt = Stmt<TypeInfo>;
pub type CheckedExpr = Expr<TypeInfo>;

#[derive(Debug, Default)]
pub struct CheckedAst {
    stmts: Vec<CheckedStmt>,
}

impl CheckedAst {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[CheckedStmt] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<CheckedStmt> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: CheckedStmt) {
        self.stmts.push(stmt);
    }

    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in &self.stmts {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Note:
// This is a massive simplification of what a real type structure would be. In a realistic case,
// `CheckedAst` would contain a vector of types (including user defined ones). Then, `CheckedInfo`
// would contain a ref-counted pointer to one of the types.

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Type {
    Stmt, // stmt only
    Integer,
    Float,
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Type::*;
        match self {
            Stmt => write!(f, "stmt"),
            Integer => write!(f, "int"),
            Float => write!(f, "float"),
        }
    }
}

/// AST meta-data after type checking.
#[derive(Debug)]
pub struct TypeInfo {
    /// type of the node
    pub type_: Type,

    /// token span of the node (typically copied over from the parsed state)
    pub tok_span: TokSpan,
}

impl TypeInfo {
    pub fn new(type_: Type, tok_span: TokSpan) -> Self {
        Self { type_, tok_span }
    }
}

impl fmt::Display for TypeInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.type_)
    }
}
