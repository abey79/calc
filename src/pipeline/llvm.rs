use crate::context::checked_ast::{CheckedBinOp, CheckedExpr, CheckedStmt, CheckedUnaryOp, Type};
use crate::data::ast::{BinOpKind, ExprKind, StmtKind, UnaryOpKind};
use crate::states::CheckedState;
use std::collections::BTreeMap;
use std::fmt;

pub(crate) fn llvm_codegen<W: fmt::Write>(input: &CheckedState, writer: &mut W) -> fmt::Result {
    let mut codegen = LlvmCodegen::new(input, writer);
    codegen.run()
}

#[derive(Debug, Clone)]
enum LlvmType {
    Builtin(Type),
    // more types here, e.g. tuple
}

impl LlvmType {
    pub fn init_val(&self) -> &'static str {
        match self {
            Self::Builtin(Type::Float) => "0.0",
            Self::Builtin(Type::Integer) => "0",
            Self::Builtin(Type::Stmt) => unreachable!(),
        }
    }
}

impl fmt::Display for LlvmType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LlvmType::Builtin(type_) => match type_ {
                Type::Integer => write!(f, "i32"),
                Type::Float => write!(f, "double"),
                Type::Stmt => unreachable!(),
            },
        }
    }
}

struct LlvmValue {
    // TODO: the representation of the value must be changed to an enum for tuple
    pub register: String,
    pub type_: LlvmType,
}

impl LlvmValue {
    fn new(register: String, type_: LlvmType) -> Self {
        Self { register, type_ }
    }
}

struct LlvmCodegen<'a, W: fmt::Write> {
    input: &'a CheckedState,
    writer: &'a mut W,

    // state
    code: Vec<String>,
    globals: BTreeMap<String, LlvmType>,
    id: usize,
}

impl<'a, W: fmt::Write> LlvmCodegen<'a, W> {
    fn new(input: &'a CheckedState, writer: &'a mut W) -> Self {
        Self {
            input,
            writer,
            code: Vec::new(),
            globals: BTreeMap::new(),
            id: 0,
        }
    }

    fn out<S: Into<String>>(&mut self, s: S) {
        self.code.push(s.into());
    }

    fn next_id(&mut self) -> usize {
        let id = self.id;
        self.id += 1;
        id
    }

    fn next_reg(&mut self) -> String {
        format!("%r{}", self.next_id())
    }

    fn run(&mut self) -> fmt::Result {
        // generate code for each statement
        for stmt in self.input.ast.stmts() {
            self.codegen_stmt(stmt)?;
        }

        // write output
        writeln!(self.writer, "declare void @_print_int(i32 %x)")?;
        writeln!(self.writer, "declare void @_print_float(double %x)")?;
        writeln!(self.writer)?;

        // declare global variables
        for (name, ltype) in &self.globals {
            writeln!(
                self.writer,
                "@{} = global {} {}",
                name,
                ltype,
                ltype.init_val()
            )?;
        }

        writeln!(self.writer)?;
        writeln!(self.writer, "define void @calc_main() {{")?;

        for line in &self.code {
            writeln!(self.writer, "    {line}")?;
        }

        writeln!(self.writer, "    ret void")?;
        writeln!(self.writer, "}}")?;

        Ok(())
    }

    fn codegen_stmt(&mut self, stmt: &CheckedStmt) -> fmt::Result {
        match &stmt.kind {
            StmtKind::Assign { name, value } => {
                let llvm_value = self.codegen_expr(value)?;
                self.out(format!(
                    "store {} {}, {}* @{}",
                    llvm_value.type_, llvm_value.register, llvm_value.type_, name
                ));
                self.globals.insert(name.to_string(), llvm_value.type_);
            }
            StmtKind::Print { expr } => {
                let llvm_value = self.codegen_expr(expr)?;

                let func = match expr.meta.type_ {
                    Type::Stmt => unreachable!("expression cannot have Stmt type"),
                    Type::Integer => "_print_int",
                    Type::Float => "_print_float",
                };

                self.out(format!(
                    "call void @{}({} {})",
                    func, llvm_value.type_, llvm_value.register
                ));
            }
            StmtKind::Expr { expr } => {
                // pointless since no possibly side effects
                self.codegen_expr(expr)?;
            }
        }
        Ok(())
    }

    fn codegen_expr(&mut self, expr: &CheckedExpr) -> Result<LlvmValue, fmt::Error> {
        match &expr.kind {
            ExprKind::Variable(name) => self.codegen_variable(name.as_ref()),
            ExprKind::UnaryOp { op, operand } => self.codegen_unary_op(op, operand),
            ExprKind::BinOp { op, left, right } => self.codegen_bin_op(op, left, right),
            ExprKind::Tuple(..) => todo!(),
            ExprKind::Integer(i) => Ok(LlvmValue::new(
                i.to_string(),
                LlvmType::Builtin(Type::Integer),
            )),
            ExprKind::Float(f) => Ok(LlvmValue::new(
                format!("{:?}", f),
                LlvmType::Builtin(Type::Float),
            )),
        }
    }

    fn codegen_variable(&mut self, name: &str) -> Result<LlvmValue, fmt::Error> {
        let reg = self.next_reg();
        let type_ = self
            .globals
            .get(name)
            .expect("type checker should have checked this")
            .clone();

        self.out(format!("{0} = load {1}, {1}* @{2}", reg, type_, name));
        Ok(LlvmValue::new(reg, type_))
    }

    fn codegen_unary_op(
        &mut self,
        op: &CheckedUnaryOp,
        operand: &CheckedExpr,
    ) -> Result<LlvmValue, fmt::Error> {
        let operand = self.codegen_expr(operand)?;

        //TODO: should match on operand.type_ when it's properly supported
        match operand.type_ {
            LlvmType::Builtin(type_) => self.codegen_unary_op_builtin(&type_, op.kind, operand),
        }
    }

    fn codegen_unary_op_builtin(
        &mut self,
        type_: &Type,
        op: UnaryOpKind,
        value: LlvmValue,
    ) -> Result<LlvmValue, fmt::Error> {
        match op {
            UnaryOpKind::Neg => {
                let reg = self.next_reg();

                let (opcode, cst) = match type_ {
                    Type::Integer => ("sub", "0"),
                    Type::Float => ("fsub", "0.0"),
                    Type::Stmt => unreachable!(),
                };

                self.out(format!(
                    "{} = {} {} {}, {}",
                    reg, opcode, value.type_, cst, value.register
                ));

                Ok(LlvmValue::new(reg, value.type_))
            }
            UnaryOpKind::Pos => Ok(value),
        }
    }

    fn codegen_bin_op(
        &mut self,
        op: &CheckedBinOp,
        left: &CheckedExpr,
        right: &CheckedExpr,
    ) -> Result<LlvmValue, fmt::Error> {
        let left = self.codegen_expr(left)?;
        let right = self.codegen_expr(right)?;

        //TODO: should match on operand.type_ when it's properly supported
        match left.type_ {
            LlvmType::Builtin(type_) => self.codegen_bin_op_builtin(&type_, op.kind, left, right),
        }
    }

    fn codegen_bin_op_builtin(
        &mut self,
        type_: &Type,
        op: BinOpKind,
        left: LlvmValue,
        right: LlvmValue,
    ) -> Result<LlvmValue, fmt::Error> {
        let opcode = match type_ {
            Type::Integer => match op {
                BinOpKind::Add => "add",
                BinOpKind::Sub => "sub",
                BinOpKind::Mul => "mul",
                BinOpKind::Div => "sdiv",
            },
            Type::Float => match op {
                BinOpKind::Add => "fadd",
                BinOpKind::Sub => "fsub",
                BinOpKind::Mul => "fmul",
                BinOpKind::Div => "fdiv",
            },
            Type::Stmt => unreachable!(),
        };

        let reg = self.next_reg();

        self.out(format!(
            "{} = {} {} {}, {}",
            reg, opcode, left.type_, left.register, right.register
        ));

        Ok(LlvmValue::new(reg, left.type_))
    }
}

#[cfg(test)]
mod test {
    use crate::pipeline::checker::check;
    use crate::pipeline::llvm::llvm_codegen;
    use crate::pipeline::parser::parse;
    use crate::pipeline::tokenizer::tokenize;
    use crate::states::InputState;

    #[test]
    fn test_llvm_codegen() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();

        let mut output = String::new();
        llvm_codegen(&checked, &mut output).unwrap();

        insta::assert_snapshot!(output);
    }
}
