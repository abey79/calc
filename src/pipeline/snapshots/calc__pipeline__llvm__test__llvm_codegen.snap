---
source: src/pipeline/llvm.rs
expression: output
---
declare void @_print_int(i32 %x)
declare void @_print_float(double %x)

@a = global double 0.0
@b = global double 0.0

define void @calc_main() {
    %r0 = fadd double 1.3, 3.2
    %r1 = fmul double %r0, 45.1
    store double %r1, double* @a
    %r2 = load double, double* @a
    %r3 = fmul double %r2, 3.2
    store double %r3, double* @b
    %r4 = mul i32 2, 3
    %r5 = add i32 1, %r4
    call void @_print_int(i32 %r5)
    ret void
}

